src/compiler/checker.ts

```diff
@@ -49508,12 +49506,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkExternalEmitHelpers(location: Node, helpers: ExternalEmitHelpers) {
-        if ((requestedExternalEmitHelpers & helpers) !== helpers && compilerOptions.importHelpers) {
-            const sourceFile = getSourceFileOfNode(location);
-            if (isEffectiveExternalModule(sourceFile, compilerOptions) && !(location.flags & NodeFlags.Ambient)) {
+        const sourceFile = getSourceFileOfNode(location);
+        if (isEffectiveExternalModule(sourceFile, compilerOptions) && !(location.flags & NodeFlags.Ambient)) {
+            const links = getNodeLinks(sourceFile);
+            links.requestedExternalEmitHelpers ??= 0 as ExternalEmitHelpers;
+            if ((links.requestedExternalEmitHelpers & helpers) !== helpers && compilerOptions.importHelpers) {
                 const helpersModule = resolveHelpersModule(sourceFile, location);
                 if (helpersModule !== unknownSymbol) {
-                    const uncheckedHelpers = helpers & ~requestedExternalEmitHelpers;
+                    const uncheckedHelpers = helpers & ~links.requestedExternalEmitHelpers;
                     for (let helper = ExternalEmitHelpers.FirstEmitHelper; helper <= ExternalEmitHelpers.LastEmitHelper; helper <<= 1) {

```

Comment:
If I’m understanding this correctly, this checking is now happening once for every source file that needs to access helpers, but the checking itself only depends on the `helpersModule`. Maybe `requestedExternalEmitHelpers` could be stored on `getSymbolLinks(helpersModule)` instead of `getNodeLinks(sourceFile)` and only happen once per instance of helpers.

However, the _error reporting_ does reference the location being passed in, and maybe that matters for your region-based diagnostics work? Although, we’re still only doing this once per requested helper function per source file, so the location of the errors can move 🤔

Reply:
> If I’m understanding this correctly, this checking is now happening once for every source file that needs to access helpers, but the checking itself only depends on the `helpersModule`. Maybe `requestedExternalEmitHelpers` could be stored on `getSymbolLinks(helpersModule)` instead of `getNodeLinks(sourceFile)` and only happen once per instance of helpers.

Good point. So `requestedExternalEmitHelperNames` should also be stored via symbol links, instead of also being a global, because now `helpersModule` could be different symbols for the same type checker.

> However, the _error reporting_ does reference the location being passed in, and maybe that matters for your region-based diagnostics work? Although, we’re still only doing this once per requested helper function per source file, so the location of the errors could be check order dependent both before and after this PR, as it currently is 🤔

Yeah, I'm planning on dealing with the error location varying in a future PR. But with this PR, we already address the problem of this error possibly jumping to a different, unopened file and never showing up for the user in the editor.

Reply:
> So then `requestedExternalEmitHelperNames` would be irrelevant.

Hmm, would it? A given instance of a helpers module might have checked `__extends` in one call, and then the same instance needs to check `__assign` in a later call, and ought to be able to skip checking `__extends` again.

Reply:
Ah, I mean `requestedExternalEmitHelperNames` and `requestedExternalEmitHelpers` should represent the same set, if `requestedExternalEmitHelpers` is now stored at the symbol level, so we only need one or the other.

Reply:
Ok, so I think I confused myself again.

Say we have two different typescript files that need the emit helper called `__awaiter`. Say those two files share the same `tslib`. Should we have an error on both typescript files saying their `tslib` is missing the emit helper called `__awaiter`? Or only on one of the files?

Reply:
I think it would be annoying on the CLI, or in any list of project-wide diagnostics, to get tslib-related errors duplicated across many/all files. I’m not sure how we might want to handle that in an editor scenario, though.

Reply:
> I think it would be annoying on the CLI, or in any list of project-wide diagnostics, to get tslib-related errors duplicated across many/all files. 

That, at least, matches what we currently do (i.e., only error on the first occurrence).

Reply:
I added test  `tslibMultipleMissingHelper.ts` that shows the specific helper checking behavior.